llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -31814,6 +31820,45 @@ static StringRef getInstrStrFromOpNo(const SmallVectorImpl<StringRef> &AsmStrs,
   return StringRef();
 }
 
+static SDValue LowerFLDEXP(SDValue Op, SelectionDAG &DAG) {
+  SDValue X = Op.getOperand(0);
+  EVT XScalarTy = X.getValueType();
+  SDValue Exp = Op.getOperand(1);
+
+  SDLoc DL(Op);
+  EVT XVT, ExpVT;
+  SDValue IID;
+  switch (Op.getSimpleValueType().SimpleTy) {
+  default:
+    return SDValue();
+  case MVT::f16:
+    X = DAG.getNode(ISD::FP_EXTEND, DL, MVT::f32, X);
```

Comment:
Can we use vscalefph if we have AVX512FP16?

Reply:
+1

Reply:
Hi, there may be risk of truncation, as the EXP operand of `FLDEXP` types `i32`,  e.g., `@llvm.ldexp.f16.i32(half, i32)` ->`@llvm.x86.avx512fp16.mask.scalef.sh(<8 x half>, <8 x half>, <8 x half>`. I didn't know how to handle the issue elegantly, so I makde an extension here.

Reply:
Could we use value tracking to check the bounds of the EXP operand?

Reply:
Value tracking seems to make things very complicated.

Reply:
I'd settle for a TODO comment for now